src/core/algorithms/fd/pfdtane/pfd_verifier/pfd_stats_calculator.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include "config/error/type.h"
+#include "config/indices/type.h"
+#include "config/error_measure/type.h"
+#include "config/indices/type.h"
+#include "model/table/column_layout_relation_data.h"
+#include "model/table/position_list_index.h"
+
+namespace algos {
+
+class PFDStatsCalculator {
+private:
+    std::shared_ptr<ColumnLayoutRelationData> relation_;
+    config::ErrorType max_fd_error_;
+    config::ErrorMeasureType error_measure_;
+
+    std::vector<model::PLI::Cluster> clusters_violating_pfd_;
+    size_t num_rows_violating_pfd_ = 0;
+    config::ErrorType error_ = 0.0;
+
+public:
+    explicit PFDStatsCalculator(std::shared_ptr<ColumnLayoutRelationData> relation,
+                                config::ErrorMeasureType measure, config::ErrorType max_fd_error)
+        : relation_(std::move(relation)), max_fd_error_(max_fd_error), error_measure_(measure) {}
+
+    void ResetState() {
+        clusters_violating_pfd_.clear();
+        num_rows_violating_pfd_ = 0;
+        error_ = 0;
+    }
+
+    bool PFDHolds() const {
+        return error_ <= max_fd_error_;
+    }
+
+    size_t GetNumViolatingClusters() const {
+        return clusters_violating_pfd_.size();
+    }
+
+    size_t GetNumViolatingRows() const {
+        return num_rows_violating_pfd_;
+    }
+
+    config::ErrorType GetError() const {
+        return error_;
+    }
+
+    std::vector<model::PLI::Cluster> const& GetViolatingClusters() const {
+        return clusters_violating_pfd_;
+    }
+
+    void CalculateStatistics(model::PositionListIndex const* x_pli,
```

Comment:
Almost entirely repeats `PFDTane::CalculatePFDError` from #396. I think we will have to split it into several functions to avoid repetition as soon as #396 is merged.